### ISSUE TITLE
Docs: publish a GPU-vs-CPU model architecture support table (no user-facing doc exists today)

### ISSUE THREAD

**Issue by alfredodeza** [paiml/aprender#3077]
## Summary

There is no user-facing documentation that says which model architectures get real GPU
inference vs. silent CPU fallback. The only source of truth is a compile-time capability
gate (`crates/aprender-serve/src/capability.rs`) that a user would have to read Rust to
discover. #3075 showed the cost of this: a user with an RTX 4090 assumed a quantized
Phi-3 GGUF would run on GPU — it can't, on any quantization, and nothing in `--help`,
the CLI output, or `docs/` says so up front.

`grep`ing `docs/` and `crates/*/docs` for any existing support matrix ("GPU support
matrix", "supported architectures", etc.) turns up nothing that documents this for users.

## Ask

Add a short, maintained doc (e.g. `docs/gpu-model-support.md`, linked from the main
README/CLAUDE.md GPU section) with a small table like:

| Architecture (`general.architecture`) | GPU (CUDA) | CPU | Recommendation |
|---|---|---|---|
| llama / llama3 | ✅ | ✅ | any quantization |
| qwen2 / qwen2.5 | ✅ | ✅ | any quantization |
| qwen3 | ✅ | ✅ | any quantization |
| qwen3-moe / qwen3.5-moe | ✅ (dedicated MoE path) | ✅ | any quantization |
| mistral | ✅ | ✅ | any quantization |
| gemma (v1) | ✅ | ✅ | any quantization |
| gemma2 / gemma3 | ❌ (attn/final-logit softcap not on GPU) | ✅ | CPU only — quantization doesn't change this |
| phi2 | ❌ (LayerNorm + GELU-MLP not on GPU) | ✅ | CPU only — use a quantized GGUF (Q4_K_M) for speed, not fp16 |
| phi / phi3 | ❌ (LayerNorm not on GPU) | ✅ | CPU only — use a quantized GGUF (Q4_K_M) for speed, not fp16 |
| gpt2 | ❌ | ✅ | CPU only |
| bert / t5 / whisper | ❌ | ✅ (unverified perf) | CPU only |
| mamba / rwkv | ⚠️ unverified | ⚠️ unverified | recurrent/SSM archs — not clear the capability gate (which only models transformer op mismatches) is even the right check here; needs explicit verification, not just a `RequiredOp` diff |

The exact rows should be **derived**, not hand-copied from this issue — see Verification
below — since the fallback constraints table drifts (this repo's CLAUDE.md convention:
derive the numbers, don't quote them).

Also add the general rule in prose, since it's the part that actually would have
prevented #3075's confusion: **GPU eligibility is decided by architecture alone
(`ArchConstraints`), never by quantization type/qtype.** Quantizing a GPU-ineligible
model only speeds up its CPU fallback (via SIMD-fused quantized matmul kernels); it
never makes it GPU-eligible.

## Where the source-of-truth data lives (for whoever writes this)

- Per-architecture ops required: `crates/aprender-serve/src/gguf/arch_constraints_fallback.rs`
  (`from_architecture_generated`) — or the generated YAML-driven version when
  `contracts/arch-constraints-v1.yaml` is present at build time.
- What the GPU backend actually supports: `crates/aprender-serve/src/capability.rs::gpu_supported_ops()`.
- Version-gated extra requirements not visible from base constraints (e.g. Gemma2/3
  softcap): `crates/aprender-serve/src/capability.rs::required_ops_for_model` /
  `arch_needs_softcap_postnorm`.
- MoE architectures use a separate dedicated GPU path with their own parity gate
  (`crates/aprender-serve/src/gguf/cuda/mod.rs`, `M-GPU-MOE-1.3`) rather than the generic
  dense capability check — worth a callout in the doc rather than lumping MoE in with
  the plain per-op diff.

## Suggested follow-up (optional, out of scope for the doc itself)

Consider a `pv`/`pmat`-style generator (or a `#[test]` in the same spirit as
`crates/aprender-core/tests/readme_contract.rs`) that renders this table straight from
`gpu_supported_ops()` × every arch in `arch_constraints_fallback.rs`, so it can't drift
silently the way the rest of this file's numbers historically have. Not required to
close this issue — a hand-written table with a "last verified against commit X" note is
an acceptable v1.

## Verification

Whoever picks this up should re-derive every row against current `main`, not copy the
table above verbatim — it was hand-derived from the two files listed and has not been
cross-checked against the YAML-generated variant of `arch_constraints`
(`contracts/arch-constraints-v1.yaml`, when present) or against an actual GPU run per
architecture.

## Related

- #3075 (Phi LayerNorm GPU gap — the concrete case that exposed the missing doc)
- #3076 (F16/BF16 CPU matmul has no SIMD path — the other half of what made the original
  report look like a hang)

---

## Raised to P1 — a user hit this today, and the self-diagnosis it produces is "my models must be wrong"

Reported 2026-09-20 by Alfredo: *"I am trying so far and can't get it to work. Have you tried it with a specific Qwen model? It is entirely possible that the quantized models I have are not exactly what apr will work with."*

That is the exact failure this issue predicts: with no published table, a user who is outside the supported set concludes **their files are wrong**. They are not. The set is narrow, it is undocumented, and nothing in the product tells them where its edge is.

### The two facts a user cannot currently discover

**1. The GPU quant whitelist is exact, and the IQ family is entirely outside it.** `crates/aprender-serve/src/gguf/dtype.rs:46`:

```rust
pub(crate) fn gpu_unsupported_quant_qtype(qtype: u32) -> bool {
    !matches!(qtype, 0 | 2 | 3 | 6 | 8 | 12 | 13 | 14)
}
```

GPU-eligible is exactly **F32, Q4_0, Q4_1, Q5_0, Q8_0, Q4_K, Q5_K, Q6_K**. Gated to CPU: F16, Q5_1, Q8_1, Q2_K, Q3_K, Q8_K, **every IQ\* type**, BF16. The Unsloth `UD-IQ2_XXS` / `UD-IQ4_XS` builds — among the most commonly downloaded Qwen quants — are CPU-only here, by design, silently.

**2. Every model we have validated is Q4_K_M.** `contracts/model-capability-ladder-v1.yaml` carries eight rungs, all `Q4_K_M`: qwen2-1.5b, qwen3-1.7b, qwen3-8b, qwen35 0.8b/2b/4b/9b/27b. The 20/20 CPU+CUDA result behind the 0.68.2 release is a **Q4_K_M** result. We have never published that qualifier.

So the honest user-facing statement today is: *Q4_K_M is the proven path; six other quant types reach the GPU; IQ runs CPU-only.* No document says any of it.

### done_when

1. A table in the README (or a page the README links) with two axes — **architecture × quant type** — and three states per cell: GPU, CPU-only, unsupported. Generated from `gpu_unsupported_quant_qtype` and the ladder contract, not hand-typed, so it cannot drift from the whitelist it describes.
2. The table states which cells are **measured** (the eight ladder rungs, with their sha256) versus merely **routable**. "The kernel exists" and "we ran it and checked the logits" are different claims and a user betting a download on it needs to know which they are getting.
3. `apr qa` / `apr inspect` names the quant type **and its routing** on any model it reads, so the answer is one command away without reading docs at all. A user holding an IQ4_XS file should be told it is CPU-only by the tool, not deduce it from a slowdown.
4. The refusal path says the same thing: where a quant is gated, the message names the quant and the reason, not the model.

Refs #3080, #3091, #3541, #3542.

**Comment by noahgift** [comment]
Train 0.68.0 because this is Alfredo's non-Qwen3.5 ticket. Waiting on docs update.

**Comment by noahgift** [comment]
Plan (0.67.0 orchestrator, supersedes the one-line triage note above): **0.68.0** (leaves 2026-09-15T06:00Z). The table is DERIVED, not authored: generated from `arch_constraints_fallback.rs` + the per-backend `gpu_supported_ops()` capability rows into docs/ (and the README's model section) by a script, with a readme_contract-style drift test so it cannot silently rot; the qwen35 row is corrected first by the #3091 honest-refusal PR (hybrid Gated DeltaNet, unsupported on both backends today), the Phi row by #3075.

**Comment by noahgift** [comment]
Cross-ref: **#3945** proposes generating this table rather than maintaining it — rendering it from the enumerated `(qtype × op × backend)` conformance matrix derived from #3421's `quant_type_traits()` registry.

The reason to prefer generation here specifically: a hand-kept support table can drift from what the code permits, and we measured exactly that this week one layer down — `dtype.rs`'s own doc comment listed ten GPU-eligible quant types while its expression permitted thirteen, wrong in both directions. A published user-facing table has the same failure mode with a worse blast radius. If it is rendered from the registry, `--quality`-style drift cannot happen.

**Comment by noahgift** [comment]
milestone → 0.74.0 per docs/roadmaps/milestone-triage-2026-09-23.md (#4024), operator approval 2026-09-23 (relayed by aprender-cf). Reason: architecture support table, generated from TRAITS with #3443: Any Model (#4001)
